swarm/src/lib.rs

```diff
@@ -1817,6 +1835,34 @@ fn p2p_addr(peer: Option<PeerId>, addr: Multiaddr) -> Result<Multiaddr, Multiadd
     Ok(addr.with(multiaddr::Protocol::P2p(peer)))
 }
 
+fn any_with_same_transport(
```

Comment:
Great! Now we just need to use it above as well! :D

Reply:
Plus we need some tests! You should be able to find plenty of addresses scattered over https://github.com/libp2p/specs.

Now, when thinking of some testcases, I realised that our current implementation is probably wrong in several ways but that is okay. We can fix that as we implement the test cases. I'd suggest you write the tests first and then we go about fixing the implementation.

- We can listen on IPv4 and IPv6 selectively so we should consider those as being separate: Listening on IPv4 and dialing IPv6 means `PortUse::New`.
- There can be many addresses behind `/dnsaddr`, meaning we should probably always allocate a new port here. This is an unfortunate interaction as DNS is resolved in the transport layer and not up here so we don't really know.
- Same thing for `/dns`, we don't know what is behind it so we should always say `PortUse::New`.
- We should test that same IP address but TCP vs QUIC vs QUICv1 vs WebRTC yields us `PortUse::New`.
- We should test that different IP address but both are TCP should give us `PortUse::Existing`.
- Same for QUIC, QUICv1 & WebRTC-direct

Arguable, the test cases for IP4/6 combined with TCP and QUIC(v1) are the most important. Those are the ones we support hole-punching for so we should test that properly.

Reply:
I'm currently writing the test cases. Quick question: Should we just include ip4 and ip6 in the list that we don't filter out and transform dns4 into ip4 as well as dns6 into ip6.
Instead of defaulting to New for dns and dnsaddr we can just look if there exists a listener if we transform dnsaddr/dns into ip4 and ip6. If there exists a listener for both we can reuse too. However this will increase complexity.